Static Pointer

adamemerson opened the pull request:
Men dream of many things.

Some dream of peace.
Some dream of wealth.
Some dream of love.

Some dream of being able to use polymorphism without allocating memory on the heap.

These last men are best, as far as this PR is concerned.

mattbenjamin commented:
@adamemerson much obscure c++ foo here, but the mechanism looks very good

mattbenjamin commented on src/test/common/test_static_ptr.cc:
```diff
@@ -0,0 +1,261 @@
+  int call(int n) override { return n + val; }
+};
+
+TEST(StaticPtr, EmptyCreation) {
```
thank you--my main use case

mattbenjamin commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,395 @@
+    new (p2) T(std::move(*me));
+    break;
+
+
```
spc

mattbenjamin commented on src/common/static_ptr.h:
````diff
@@ -0,0 +1,395 @@
+// I *recommend* having a size constant in header files (or perhaps a
+// using declaration, e.g.
+// ```
+// using StaticFoo = static_ptr<Foo, sizeof(Blah)>`
````
ok

mattbenjamin commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,395 @@
+  // construction because in-place construction renders it
+  // unnecessary. Also i doesn't fit the pointer idiom as well.
+  //
+  template<typename T, typename... Args>
```
wow, I actually understand this

mattbenjamin commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,395 @@
+// *_pointer_cast overloads for a move-only unique pointer. It's a
+// nice idiom. Having to release and reconstruct is obnoxious.
+//
+template<typename U, std::size_t Z, typename T, std::size_t S>
```
ok

mattbenjamin commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,395 @@
+// I'm not sure if anyone will ever use this. I can imagine situations
+// where they might. It works, though!
+//
+template<typename U, std::size_t Z, typename T, std::size_t S>
```
scary

mattbenjamin commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,395 @@
+// enough, we copy it over.
+//
+// I follow cast semantics. Since this is a pointer-like type, it
+// returns a null value rather than throwing.
```
what is this used for?

adamemerson replied:
If you have a static_ptr<Base, 4096> in some part of the code and you have a static_ptr<Base, 256> in some other part of the code.

(Maybe the large derivative are only used on some call paths and you don't want to bloat every structure everywhere)

If you know that the actual thing inside the 4096-byte static pointer is only 64 bytes long, resize_pointer_cast will check that it actually IS small enough to fit and copy it. (If it's not small enough to fit, it will just return a null value, similar to dynamic_cast.)

In normal use we just check the declared size at compile time and refuse to copy/move from a bigger static_ptr into a smaller one. This is an escape hatch where you can request to have it checked and done.

adamemerson commented:
@cbodley All right, if the supplied type is nothrow move/copy constructible, so are we.

cbodley commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,409 @@
+      operate(_mem::op::copy, const_cast<void*>(static_cast<const void*>(&rhs.buf)), &buf);
+    }
+    operate = rhs.operate;
+    operate(_mem::op::move, &rhs.buf, &buf);
```
this does both copy and move. i'm guessing you just want the move inside the if block

cbodley commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,409 @@
+  copy, move, destroy, size
+};
+template<typename T>
+static std::size_t op_fun(op oper, void* p1, void* p2) noexcept
```
i don't think it's safe to make this `noexcept` unless T is both nothrow-move- and nothrow-copy-constructible. maybe the unit tests should throw some exceptions for those to prove that we won't terminate

cbodley commented on src/common/static_ptr.h:
```diff
@@ -0,0 +1,409 @@
+    // or copy methods. Working around this is a pain, since we'd need
+    // four operator functions and a top-level class to
+    // provide/withhold copy/move operations as appropriate.
+    new (p2) T(*me);
```
i'm not sure it's reasonable to require move+copy support from all of our types. could we explore some workarounds?

mattbenjamin replied:
I'd like to proceed quickly to a working model.  If the current generality is the problem, I'm in favor of reducing scope.

adamemerson replied:
@cbodley I'd rather just fix the two safety issues you pointed out (doing that now) and get this version in. The more general version can be done with a template switch+inheritance, but I'd rather do that in a second patch just so Matt can have at it and we can get some feedback before complicating things.

adamemerson commented:
All right!

Fixed the stuff Casey pointed out, added a test to make sure exceptions work and fixed a problem with default copy constructors.

@mattbenjamin I think we're ready to merge.
